dcos/util.py

```diff
@@ -98,6 +100,63 @@ def remove_path_on_error(path):
         raise
 
 
+@contextlib.contextmanager
+def silent_outputs():
+    """A context manager for suppressing stdout / stderr, it sets their file
+    descriptors to os.devnull and then restores them.
+
+    This is helpful for inhibiting output from C code or subprocesses, as
+    replacing sys.stdout and sys.stderr wouldn't be enough.
+    """
+
+    libc = ctypes.CDLL(None)
+    c_stdout = ctypes.c_void_p.in_dll(libc, 'stdout')
+    c_stderr = ctypes.c_void_p.in_dll(libc, 'stderr')
+
+    # Original fds stdout/stderr point to. Usually 1 and 2 on POSIX systems.
+    original_stdout_fd = sys.stdout.fileno()
+    original_stderr_fd = sys.stderr.fileno()
+
+    def _redirect_outputs(to_fd):
+        """Redirect stdout/stderr to the given file descriptor."""
+        # Flush the C-level stdout / stderr buffers
```

Comment:
"the C-level stdout / stderr buffers" is not well-defined in my mind... what is that supposed to be? Why is `libc` interaction required here?

I feel like I have messed quite a bit with the CPython I/O system on various different platforms and I never needed to import `ctypes` for that.

Why is e.g. `sys.stdout.flush()` not sufficient?

Reply:
I don't think this is needed in the specific case we're trying to solve (output from a subprocess), but this was inspired by https://eli.thegreenplace.net/2015/redirecting-all-kinds-of-stdout-in-python/, where they say this is needed in case we're running C code (eg. a python module implemented as a C extension)  :
 
```With this information in hand, it should be easy to understand what stdout actually is for a C program. stdout is a global FILE object kept for us by the C library, and it buffers output to file descriptor number 1. Calls to functions like printf and puts add data into this buffer. fflush forces its flushing to the file descriptor, and so on.```

So even though we flush `sys.stdout`, there might still be some buffered output from C code.